// kpc/@css/components/index.js
'use strict';

exports.__esModule = true;
exports.localize = exports._$ = exports.version = exports.Upload = exports.Transfer = exports.Tooltip = exports.Tag = exports.Tabs = exports.TableColumn = exports.Table = exports.Tab = exports.Switch = exports.Steps = exports.Step = exports.Spinner = exports.Slider = exports.Select = exports.ScrollSelect = exports.Row = exports.Radio = exports.Progress = exports.Pagination = exports.OptionGroup = exports.Option = exports.MoveWrapper = exports.Message = exports.Link = exports.Input = exports.FormItem = exports.Form = exports.Editable = exports.DropdownMenu = exports.DropdownItem = exports.Dropdown = exports.Dialog = exports.Datepicker = exports.CollapseItem = exports.Collapse = exports.Col = exports.Checkbox = exports.Cascader = exports.ButtonGroup = exports.Button = exports.BreadcrumbItem = exports.Breadcrumb = exports.Badge = exports.App = undefined;
exports.install = install;

var _utils = require('./utils');

var _app = require('./app');

var _badge = require('./badge');

var _breadcrumb = require('./breadcrumb');

var _button = require('./button');

var _cascader = require('./cascader');

var _checkbox = require('./checkbox');

var _collapse = require('./collapse');

var _datepicker = require('./datepicker');

var _dialog = require('./dialog');

var _dropdown = require('./dropdown');

var _editable = require('./editable');

var _form = require('./form');

var _input = require('./input');

var _link = require('./link');

var _message = require('./message');

var _moveWrapper = require('./moveWrapper');

var _pagination = require('./pagination');

var _progress = require('./progress');

var _radio = require('./radio');

var _grid = require('./grid');

var _scrollSelect = require('./scrollSelect');

var _select = require('./select');

var _slider = require('./slider');

var _spinner = require('./spinner');

var _steps = require('./steps');

var _switch = require('./switch');

var _table = require('./table');

var _tabs = require('./tabs');

var _tag = require('./tag');

var _tooltip = require('./tooltip');

var _transfer = require('./transfer');

var _upload = require('./upload');

/* generate start */

exports.App = _app.App;
exports.Badge = _badge.Badge;
exports.Breadcrumb = _breadcrumb.Breadcrumb;
exports.BreadcrumbItem = _breadcrumb.BreadcrumbItem;
exports.Button = _button.Button;
exports.ButtonGroup = _button.ButtonGroup;
exports.Cascader = _cascader.Cascader;
exports.Checkbox = _checkbox.Checkbox;
exports.Col = _grid.Col;
exports.Collapse = _collapse.Collapse;
exports.CollapseItem = _collapse.CollapseItem;
exports.Datepicker = _datepicker.Datepicker;
exports.Dialog = _dialog.Dialog;
exports.Dropdown = _dropdown.Dropdown;
exports.DropdownItem = _dropdown.DropdownItem;
exports.DropdownMenu = _dropdown.DropdownMenu;
exports.Editable = _editable.Editable;
exports.Form = _form.Form;
exports.FormItem = _form.FormItem;
exports.Input = _input.Input;
exports.Link = _link.Link;
exports.Message = _message.Message;
exports.MoveWrapper = _moveWrapper.MoveWrapper;
exports.Option = _select.Option;
exports.OptionGroup = _select.OptionGroup;
exports.Pagination = _pagination.Pagination;
exports.Progress = _progress.Progress;
exports.Radio = _radio.Radio;
exports.Row = _grid.Row;
exports.ScrollSelect = _scrollSelect.ScrollSelect;
exports.Select = _select.Select;
exports.Slider = _slider.Slider;
exports.Spinner = _spinner.Spinner;
exports.Step = _steps.Step;
exports.Steps = _steps.Steps;
exports.Switch = _switch.Switch;
exports.Tab = _tabs.Tab;
exports.Table = _table.Table;
exports.TableColumn = _table.TableColumn;
exports.Tabs = _tabs.Tabs;
exports.Tag = _tag.Tag;
exports.Tooltip = _tooltip.Tooltip;
exports.Transfer = _transfer.Transfer;
exports.Upload = _upload.Upload;
var version = exports.version = '0.4.3';

/* generate end */

function install(Vue) {
    if (install.installed) return;

    if (Vue) {
        for (var key in exports) {
            var code = key.charCodeAt(0);
            // the key which the first char is uppercase is a component
            if (code >= 65 && code <= 90) {
                Vue.component(key, exports[key]);
            }
        }
    } else {
        for (var _key in exports) {
            var _code = _key.charCodeAt(0);
            // the key which the first char is uppercase is a component
            if (_code >= 65 && _code <= 90) {
                global['K' + _key] = exports[_key];
            }
        }
    }
}

exports._$ = _utils._$;
exports.localize = _utils.localize;
