// kpc/@stylus/components/utils.js
'use strict';

exports.__esModule = true;
exports.expandAnimationCallbacks = exports.isFunction = exports.isObject = exports.isNullOrUndefined = exports.get = undefined;

var _typeof2 = require('babel-runtime/helpers/typeof');

var _typeof3 = _interopRequireDefault(_typeof2);

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _keys = require('babel-runtime/core-js/object/keys');

var _keys2 = _interopRequireDefault(_keys);

exports.addStyle = addStyle;
exports.kebabCase = kebabCase;
exports.proxyEvent = proxyEvent;
exports.isStringOrNumber = isStringOrNumber;
exports.isTextVNode = isTextVNode;
exports.isStringOrNumberNotEmpty = isStringOrNumberNotEmpty;
exports.getTextByChildren = getTextByChildren;
exports.findParentComponent = findParentComponent;
exports.strPad = strPad;
exports.range = range;
exports.selectInput = selectInput;
exports._$ = _$;
exports.localize = localize;
exports.getTransition = getTransition;
exports.mapChildren = mapChildren;

var _intact = require('intact');

var _intact2 = _interopRequireDefault(_intact);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var utils = _intact2.default.utils;
var get = utils.get,
    isNullOrUndefined = utils.isNullOrUndefined,
    isObject = utils.isObject,
    isFunction = utils.isFunction;
exports.get = get;
exports.isNullOrUndefined = isNullOrUndefined;
exports.isObject = isObject;
exports.isFunction = isFunction;
function addStyle(style, extra) {
    if (!style) return extra;
    if (!extra) return style;
    if (typeof style === 'string') {
        extra = (0, _keys2.default)(extra).map(function (key) {
            return kebabCase(key) + ': ' + extra[key];
        }).join(';');
        style = '' + style + (!style || /;\s*$/.test(style) ? '' : ';') + extra;
    } else {
        style = (0, _extends3.default)({}, style, extra);
    }

    return style;
}

var _cache = {};
var uppercasePattern = /[A-Z]/g;
function kebabCase(word) {
    if (!_cache[word]) {
        _cache[word] = word.replace(uppercasePattern, function (item) {
            return '-' + item.toLowerCase();
        });
    }
    return _cache[word];
}

function proxyEvent(component, name) {
    return function (e) {
        return component.trigger(name, e);
    };
}

function isStringOrNumber(o) {
    var type = typeof o === 'undefined' ? 'undefined' : (0, _typeof3.default)(o);
    return type === 'string' || type === 'number';
}

function isTextVNode(o) {
    return isStringOrNumber(o) || o.type === 1;
}

function isStringOrNumberNotEmpty(o) {
    var type = typeof o === 'undefined' ? 'undefined' : (0, _typeof3.default)(o);
    return type === 'string' && !/^\s+$/.test(o) || type === 'number';
}

function getTextByChildren(children) {
    var ret = '';
    if (Array.isArray(children)) {
        children.forEach(function (vNode) {
            if (isTextVNode(vNode)) {
                ret += vNode.children;
            }
        });
    } else if (isStringOrNumber(children)) {
        ret += children;
    }

    return ret;
}

function findParentComponent(Component, instance, isUntil) {
    var ret = void 0;
    var parent = instance.parentVNode;
    while (parent) {
        if (parent.tag === Component) {
            ret = parent.children;
            if (isUntil) break;
        }
        parent = parent.parentVNode;
    }
    return ret;
}

function strPad(str, length, pad) {
    str = str.toString();
    pad = pad === undefined ? '0' : pad;
    var l = str.length;
    if (l < length) {
        str = new Array(length - l + 1).join(pad) + str;
    }
    return str;
}

function range(start, end) {
    return Array.apply(null, { length: end - start + 1 }).map(function (v, i) {
        return i + start;
    });
}

function selectInput(input) {
    if (input.select) {
        input.select();
    } else if (input.setSelectionRange) {
        // mobile safari
        input.focus();
        input.setSelectionRange(0, input.value.length);
    }
}

var i18n = {};
var valueRegexp = /\{([^\}\s]+)\}/g;
function _$(key, data) {
    var value = get(i18n, key);
    if (isNullOrUndefined(value)) {
        value = key;
    }

    if (data) {
        value = value.replace(valueRegexp, function (nouse, variable) {
            var suffix = void 0;
            var index = variable.indexOf(':');
            if (index > 0) {
                suffix = variable.substr(0, index);
                suffix = suffix.split('|');
                variable = variable.substr(index + 1);
                variable = get(data, variable);
                if (variable > 1) {
                    return suffix.length > 1 ? suffix[1] : suffix[0];
                } else {
                    return suffix.length > 1 ? suffix[0] : '';
                }
            } else {
                variable = get(data, variable);
                return isNullOrUndefined(variable) ? nouse : variable;
            }
        });
    }

    return value;
}
function localize(data) {
    if (!isObject(data)) {
        throw new Error('expect a object but got ' + data);
    }
    i18n = data;
}

function getTransition(feedback) {
    var horizontal = feedback.horizontal;
    var vertical = feedback.vertical;
    if (feedback.important === 'horizontal') {
        if (horizontal === 'left') {
            return 'slideright';
        } else if (horizontal === 'right') {
            return 'slideleft';
        } else if (vertical === 'bottom') {
            return 'slideup';
        } else if (vertical === 'top') {
            return 'slidedown';
        }
    } else {
        if (vertical === 'bottom') {
            return 'slideup';
        } else if (vertical === 'top') {
            return 'slidedown';
        } else if (horizontal === 'left') {
            return 'slideright';
        } else if (horizontal === 'right') {
            return 'slideleft';
        }
    }
    return 'slidedown';
}

function mapChildren(children, callback) {
    var vNodes = [];
    function map(children) {
        if (!children) return;
        if (!Array.isArray(children)) {
            return vNodes.push(callback(children));
        }
        children.forEach(function (vNode) {
            if (Array.isArray(vNode)) {
                map(vNode);
            } else if (vNode) {
                vNodes.push(callback(vNode));
            }
        });
    }

    map(children);

    return vNodes;
}

var setHeight = function setHeight(el) {
    return el.style.height = el.children[0].offsetHeight + 'px';
};
var expandAnimationCallbacks = exports.expandAnimationCallbacks = {
    'ev-a:leaveStart': setHeight,
    'ev-a:enterStart': setHeight,
    'ev-a:enterEnd': function evAEnterEnd(el) {
        return el.style.height = '';
    }
};
