// kpc/@css/components/collapse/index.js
'use strict';

exports.__esModule = true;
exports.CollapseItem = exports.Collapse = exports.default = undefined;

var _getOwnPropertyDescriptor = require('babel-runtime/core-js/object/get-own-property-descriptor');

var _getOwnPropertyDescriptor2 = _interopRequireDefault(_getOwnPropertyDescriptor);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _dec, _desc, _value2, _class, _init, _class2, _temp;

var _intact = require('intact');

var _intact2 = _interopRequireDefault(_intact);

var _index = require('./index.vdt');

var _index2 = _interopRequireDefault(_index);

var _item = require('./item');

var _item2 = _interopRequireDefault(_item);

require('../../styles/kpc.css');

require('./index.css');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _applyDecoratedDescriptor(target, property, decorators, descriptor, context) {
    var desc = {};
    Object['ke' + 'ys'](descriptor).forEach(function (key) {
        desc[key] = descriptor[key];
    });
    desc.enumerable = !!desc.enumerable;
    desc.configurable = !!desc.configurable;

    if ('value' in desc || desc.initializer) {
        desc.writable = true;
    }

    desc = decorators.slice().reverse().reduce(function (desc, decorator) {
        return decorator(target, property, desc) || desc;
    }, desc);

    if (context && desc.initializer !== void 0) {
        desc.value = desc.initializer ? desc.initializer.call(context) : void 0;
        desc.initializer = undefined;
    }

    if (desc.initializer === void 0) {
        Object['define' + 'Property'](target, property, desc);
        desc = null;
    }

    return desc;
}

var Collapse = (_dec = _intact2.default.template(), (_class = (_temp = _class2 = function (_Intact) {
    (0, _inherits3.default)(Collapse, _Intact);

    function Collapse() {
        (0, _classCallCheck3.default)(this, Collapse);
        return (0, _possibleConstructorReturn3.default)(this, _Intact.apply(this, arguments));
    }

    Collapse.prototype.defaults = function defaults() {
        return {
            value: undefined,
            accordion: false,
            arrow: 'right', // 'right' | 'left'
            noBorder: false
        };
    };

    Collapse.prototype._changeValue = function _changeValue(v) {
        var _get = this.get(),
            value = _get.value,
            accordion = _get.accordion;

        if (!Array.isArray(value)) {
            this.set('value', [v]);
        } else {
            var index = value.indexOf(v);
            var _value = value.slice(0);
            if (~index) {
                _value.splice(index, 1);
            } else if (accordion) {
                _value = [v];
            } else {
                _value.push(v);
            }
            this.set('value', _value);
        }
    };

    Collapse.prototype._isActive = function _isActive(v) {
        var _get2 = this.get(),
            value = _get2.value;

        if (!Array.isArray(value)) return false;
        return ~value.indexOf(v);
    };

    return Collapse;
}(_intact2.default), _class2.template = _index2.default, _class2.propTypes = {
    accordion: Boolean,
    noBorder: Boolean
}, _temp), (_applyDecoratedDescriptor(_class, 'template', [_dec], (_init = (0, _getOwnPropertyDescriptor2.default)(_class, 'template'), _init = _init ? _init.value : undefined, {
    enumerable: true,
    configurable: true,
    writable: true,
    initializer: function initializer() {
        return _init;
    }
}), _class)), _class));
exports.default = Collapse;
exports.Collapse = Collapse;
exports.CollapseItem = _item2.default;
