// kpc/@stylus/components/select/index.vdt.js
'use strict';

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _input = require('../input');

var _input2 = _interopRequireDefault(_input);

var _dropdown = require('../dropdown');

var _dropdown2 = _interopRequireDefault(_dropdown);

var _utils = require('../utils');

var _option = require('./option');

var _option2 = _interopRequireDefault(_option);

var _group = require('./group');

var _group2 = _interopRequireDefault(_group);

var _tabs = require('../tabs');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

module.exports = function (obj, _Vdt, blocks, $callee) {
    var _classNameObj,
        _this = this;

    _Vdt || (_Vdt = Vdt);
    obj || (obj = {});
    blocks || (blocks = {});
    var h = _Vdt.miss.h,
        hc = _Vdt.miss.hc,
        hu = _Vdt.miss.hu,
        widgets = this && this.widgets || {},
        _blocks = {},
        __blocks = {},
        __u = _Vdt.utils,
        extend = __u.extend,
        _e = __u.error,
        _className = __u.className,
        __o = __u.Options,
        _getModel = __o.getModel,
        _setModel = __o.setModel,
        _setCheckboxModel = __u.setCheckboxModel,
        _detectCheckboxChecked = __u.detectCheckboxChecked,
        _setSelectModel = __u.setSelectModel,
        self = this.data,
        $this = this,
        scope = obj,
        Animate = self && self.Animate,
        parent = ($callee || {})._super;

    var _self$get = self.get(),
        className = _self$get.className,
        clearable = _self$get.clearable,
        data = _self$get.data,
        value = _self$get.value,
        multiple = _self$get.multiple,
        disabled = _self$get.disabled,
        filterable = _self$get.filterable,
        keywords = _self$get.keywords,
        _show = _self$get._show,
        placeholder = _self$get.placeholder,
        size = _self$get.size,
        style = _self$get.style,
        name = _self$get.name,
        children = _self$get.children,
        fluid = _self$get.fluid,
        width = _self$get.width,
        allowUnmatch = _self$get.allowUnmatch,
        card = _self$get.card;

    var _activeLabel = self.get('_activeLabel');

    var _placeholder = placeholder == null ? allowUnmatch && filterable ? (0, _utils._$)('请输入或选择') : (0, _utils._$)('请选择') : placeholder;

    var classNameObj = (_classNameObj = {
        'k-select': true
    }, _classNameObj[className] = className, _classNameObj[scope.className] = scope.className, _classNameObj['k-disabled'] = disabled, _classNameObj['k-show'] = _show, _classNameObj['k-clearable'] = clearable, _classNameObj['k-' + size] = size !== 'default', _classNameObj['k-fluid'] = fluid, _classNameObj);

    var hasValue = value != null && (!multiple && value !== '' || multiple && value.length);
    var isGroup = Array.isArray(self.get('data.0.data'));

    var labelObj = scope.labelObj || {
        // only show value as label when allowUnmatch is true, #40
        label: allowUnmatch && (0, _utils.isStringOrNumber)(value) ? value : null,
        labels: []
    };
    var groupIndex = 0;

    var Options = function Options(props) {
        var data = props.data;
        var ret = [];
        data.forEach(function (item, index) {
            // deprecate text, use label instead of
            if (item.text) item.label = item.text;

            var _self$handleProps = self.handleProps(item, labelObj),
                valid = _self$handleProps.valid,
                active = _self$handleProps.active;

            if (valid) {
                ret.push(h(_dropdown.DropdownItem, { 'ev-select': function () {
                        try {
                            return [self._onSelect.bind(self, item.value)][0];
                        } catch (e) {
                            _e(e);
                        }
                    }.call(_this), 'disabled': function () {
                        try {
                            return [item.disabled][0];
                        } catch (e) {
                            _e(e);
                        }
                    }.call(_this), 'className': _className(function () {
                        try {
                            return [{ 'k-active': active }][0];
                        } catch (e) {
                            _e(e);
                        }
                    }.call(_this)), 'hideOnSelect': function () {
                        try {
                            return [!multiple][0];
                        } catch (e) {
                            _e(e);
                        }
                    }.call(_this), 'children': function () {
                        try {
                            return [item.template ? item.template(item, index) : item.label][0];
                        } catch (e) {
                            _e(e);
                        }
                    }.call(_this), '_context': $this }));

                // for card type groups, we need to highlight the parent label
                if (active && card) {
                    if (_activeLabel == null && _show) {
                        _activeLabel = groupIndex;
                        self._setActiveLabelSilent(_activeLabel);
                    }
                }
            }
        });

        return ret;
    };

    var OptionsVNodes = function OptionsVNodes(_ref) {
        var children = _ref.children;

        return __u.map(children ? Array.isArray(children) ? children : [children] : children, function (vNode) {
            if (vNode.tag === _option2.default) {
                var props = vNode.props;
                props = (0, _extends3.default)({}, props, {
                    'ev-select': self._onSelect.bind(self, props.value),
                    hideOnSelect: !multiple,
                    label: props.label || (0, _utils.getTextByChildren)(props.children),
                    children: props.children || props.label
                });

                var _self$handleProps2 = self.handleProps(props, labelObj),
                    active = _self$handleProps2.active,
                    valid = _self$handleProps2.valid;

                if (!valid) {
                    return hc('');
                } else {
                    var _className2;

                    props.className = _className((_className2 = {}, _className2[props.className] = props.className, _className2['k-active'] = active, _className2));
                }

                // for card type groups, we need to highlight the parent label
                if (active && card) {
                    if (_activeLabel == null && _show) {
                        _activeLabel = groupIndex;
                        self._setActiveLabelSilent(_activeLabel);
                    }
                }

                // if the multiple select option is active
                // add a √ icon
                if (active && multiple) {
                    props.children = [props.children, h('i', null, null, 'k-checkmark ion-ios-checkmark-empty')];
                }

                // return the cloned vNode
                return h(_option2.default, props);
            } else if (vNode.tag === _group2.default) {
                var _props = vNode.props;
                _props = (0, _extends3.default)({}, _props, {
                    children: h(OptionsVNodes, { 'children': function () {
                            try {
                                return [_props.children][0];
                            } catch (e) {
                                _e(e);
                            }
                        }.call(_this), '_context': $this })
                });
                groupIndex++;
                return h(_group2.default, _props);
            }

            return vNode;
        });
    };

    var labelList = [];

    var Menu = scope.Menu || h(_dropdown.DropdownMenu, { 'ev-$changed:show': function () {
            try {
                return [self._position][0];
            } catch (e) {
                _e(e);
            }
        }.call(this), 'className': _className(function () {
            try {
                return [{
                    "k-select-dropdown": true,
                    "k-card": card
                }][0];
            } catch (e) {
                _e(e);
            }
        }.call(this)), 'ev-$change:show': function () {
            try {
                return [self._onChangeShow][0];
            } catch (e) {
                _e(e);
            }
        }.call(this), 'children': function () {
            var _this2 = this;

            try {
                return [function () {
                    var _children = [];
                    if (!isGroup) {
                        _children.push(h(Options, { 'data': function () {
                                try {
                                    return [data][0];
                                } catch (e) {
                                    _e(e);
                                }
                            }.call(_this2), 'children': null, '_context': $this }));
                    } else {
                        _children.push(_Vdt.utils.map(function () {
                            try {
                                return [data][0];
                            } catch (e) {
                                _e(e);
                            }
                        }.call(_this2), function (value, key) {
                            return h(_group2.default, { 'label': function () {
                                    try {
                                        return [value.label][0];
                                    } catch (e) {
                                        _e(e);
                                    }
                                }.call(this), 'children': h(Options, { 'data': function () {
                                        try {
                                            return [value.data][0];
                                        } catch (e) {
                                            _e(e);
                                        }
                                    }.call(this), 'parentLabel': function () {
                                        try {
                                            return [value.label][0];
                                        } catch (e) {
                                            _e(e);
                                        }
                                    }.call(this), 'children': null, '_context': $this }), '_context': $this });
                        }, _this2));
                    }
                    _children.push(h(OptionsVNodes, { 'children': function () {
                            try {
                                return [children][0];
                            } catch (e) {
                                _e(e);
                            }
                        }.call(_this2), '_context': $this }));

                    // if it is card type group, we extract the label as navBar
                    if (card) {
                        var _flatten = function _flatten(children) {
                            children.forEach(function (vNode) {
                                if (Array.isArray(vNode)) {
                                    _flatten(vNode);
                                } else if (vNode) {
                                    if (vNode.tag === _group2.default) {
                                        var props = vNode.props;
                                        var label = (props._blocks || {}).label || props.label;
                                        labelList.push({ label: label });
                                        vNode.props = (0, _extends3.default)({}, props, { _card: true });

                                        // if the _activeLabel is undefined
                                        // we active the first label
                                        if (_activeLabel == null && _show) {
                                            _activeLabel = index;
                                            self._setActiveLabelSilent(index);
                                        }

                                        // only render the active options
                                        if (_activeLabel === index) {
                                            vNodes.push(vNode);
                                        }
                                        index++;
                                    } else {
                                        vNodes.push(vNode);
                                    }
                                }
                            });
                        };

                        var vNodes = [];
                        var index = 0;

                        _flatten(_children);
                        return [h(_tabs.Tabs, { 'vertical': function () {
                                try {
                                    return [true][0];
                                } catch (e) {
                                    _e(e);
                                }
                            }.call(_this2), 'type': 'border-card', 'v-model': '_activeLabel', 'children': _Vdt.utils.map(function () {
                                try {
                                    return [labelList][0];
                                } catch (e) {
                                    _e(e);
                                }
                            }.call(_this2), function (value, key) {
                                return h(_tabs.Tab, { 'value': function () {
                                        try {
                                            return [key][0];
                                        } catch (e) {
                                            _e(e);
                                        }
                                    }.call(this), 'children': ['\n                        ', function () {
                                        try {
                                            return [typeof value.label === 'function' ?
                                            // pass by block
                                            value.label(function () {}) :
                                            // pass by property
                                            value.label][0];
                                        } catch (e) {
                                            _e(e);
                                        }
                                    }.call(this), '\n                    '], '_context': $this });
                            }, _this2), '_context': $this, value: _getModel(self, '_activeLabel'), 'ev-$change:value': function ev$changeValue(__c, __n) {
                                _setModel(self, '_activeLabel', __n, $this);
                            } }), h('div', null, function () {
                            try {
                                return [vNodes][0];
                            } catch (e) {
                                _e(e);
                            }
                        }.call(_this2), 'k-groups')];
                    } else {
                        return _children;
                    }
                }()][0];
            } catch (e) {
                _e(e);
            }
        }.call(this), '_context': $this }, null, null, null, function (i) {
        widgets['menu'] = i;
    });

    // if the value is not in options, then set hasValue to false
    if (hasValue && !labelObj.label && !labelObj.labels.length) {
        hasValue = false;
        // if this is a value, but the value does not exist in options
        // we set it to empty, #41
        self._clearValue();
    }

    return h('div', { 'style': function () {
            try {
                return [width != null ? (0, _utils.addStyle)(style, { width: width + 'px' }) : style][0];
            } catch (e) {
                _e(e);
            }
        }.call(this), 'tabindex': function () {
            try {
                return [disabled ? "-1" : "0"][0];
            } catch (e) {
                _e(e);
            }
        }.call(this), 'ev-keypress': function () {
            try {
                return [self._onKeypress][0];
            } catch (e) {
                _e(e);
            }
        }.call(this), 'ev-keydown': function () {
            try {
                return [self._onKeydown][0];
            } catch (e) {
                _e(e);
            }
        }.call(this) }, h(_dropdown2.default, { 'position': function () {
            try {
                return [{ my: 'left top+8', at: 'left bottom' }][0];
            } catch (e) {
                _e(e);
            }
        }.call(this), 'trigger': 'click', 'disabled': function () {
            try {
                return [disabled][0];
            } catch (e) {
                _e(e);
            }
        }.call(this), 'children': [h('div', { 'ev-click': function () {
                try {
                    return [self._focusInput][0];
                } catch (e) {
                    _e(e);
                }
            }.call(this), 'tabindex': '-1' }, [h('input', { 'type': 'hidden', 'value': function () {
                try {
                    return [value][0];
                } catch (e) {
                    _e(e);
                }
            }.call(this), 'name': function () {
                try {
                    return [name][0];
                } catch (e) {
                    _e(e);
                }
            }.call(this) }), function () {
            try {
                return [!multiple && filterable][0];
            } catch (e) {
                _e(e);
            }
        }.call(this) ? h(_input2.default, { 'value': function () {
                try {
                    return [keywords == null ? labelObj.label : keywords][0];
                } catch (e) {
                    _e(e);
                }
            }.call(this), 'ev-input': function () {
                try {
                    return [self._onSearch][0];
                } catch (e) {
                    _e(e);
                }
            }.call(this), 'ev-blur': function () {
                try {
                    return [self._onBlur][0];
                } catch (e) {
                    _e(e);
                }
            }.call(this), 'ev-focus': function () {
                try {
                    return [self._selectInput][0];
                } catch (e) {
                    _e(e);
                }
            }.call(this), 'disabled': function () {
                try {
                    return [disabled][0];
                } catch (e) {
                    _e(e);
                }
            }.call(this), 'placeholder': function () {
                try {
                    return [_placeholder][0];
                } catch (e) {
                    _e(e);
                }
            }.call(this), 'size': function () {
                try {
                    return [size][0];
                } catch (e) {
                    _e(e);
                }
            }.call(this), 'fluid': function () {
                try {
                    return [true][0];
                } catch (e) {
                    _e(e);
                }
            }.call(this), 'children': null, '_context': $this }, null, null, null, function (i) {
            widgets['input'] = i;
        }) : function () {
            try {
                return [!filterable && !hasValue][0];
            } catch (e) {
                _e(e);
            }
        }.call(this) ? h('span', null, function () {
            try {
                return [_placeholder][0];
            } catch (e) {
                _e(e);
            }
        }.call(this), 'k-placeholder c-ellipsis') : function () {
            try {
                return [!multiple][0];
            } catch (e) {
                _e(e);
            }
        }.call(this) ? h('span', null, function () {
            try {
                return [labelObj.label][0];
            } catch (e) {
                _e(e);
            }
        }.call(this), 'k-value c-ellipsis') : h(Animate, { 'a:tag': 'span', 'a:disabled': function () {
                try {
                    return [true][0];
                } catch (e) {
                    _e(e);
                }
            }.call(this), 'className': 'k-values', 'children': [_Vdt.utils.map(function () {
                try {
                    return [labelObj.labels][0];
                } catch (e) {
                    _e(e);
                }
            }.call(this), function (item, index) {
                return h(Animate, { 'a:tag': 'span', 'className': 'k-select-tag', 'key': function () {
                        try {
                            return [value[index]][0];
                        } catch (e) {
                            _e(e);
                        }
                    }.call(this), 'a:transition': 'fade', 'children': [h('span', null, function () {
                        try {
                            return [item][0];
                        } catch (e) {
                            _e(e);
                        }
                    }.call(this), 'k-text'), h('i', { 'ev-click': function () {
                            try {
                                return [self._delete.bind(self, value[index])][0];
                            } catch (e) {
                                _e(e);
                            }
                        }.call(this) }, null, 'k-icon ion-ios-close-empty')], '_context': $this });
            }, this), function () {
                try {
                    return [filterable][0];
                } catch (e) {
                    _e(e);
                }
            }.call(this) ? h(_input2.default, { 'value': function () {
                    try {
                        return [keywords][0];
                    } catch (e) {
                        _e(e);
                    }
                }.call(this), 'ev-input': function () {
                    try {
                        return [self._onSearch][0];
                    } catch (e) {
                        _e(e);
                    }
                }.call(this), 'ev-blur': function () {
                    try {
                        return [self._onBlur][0];
                    } catch (e) {
                        _e(e);
                    }
                }.call(this), 'ev-focus': function () {
                    try {
                        return [self._onFocus][0];
                    } catch (e) {
                        _e(e);
                    }
                }.call(this), 'disabled': function () {
                    try {
                        return [disabled][0];
                    } catch (e) {
                        _e(e);
                    }
                }.call(this), 'placeholder': function () {
                    try {
                        return [!hasValue ? _placeholder : ''][0];
                    } catch (e) {
                        _e(e);
                    }
                }.call(this), 'autoWidth': function () {
                    try {
                        return [true][0];
                    } catch (e) {
                        _e(e);
                    }
                }.call(this), 'size': function () {
                    try {
                        return [size][0];
                    } catch (e) {
                        _e(e);
                    }
                }.call(this), 'children': null, '_context': $this }, null, null, null, function (i) {
                widgets['input'] = i;
            }) : undefined], '_context': $this }), h('span', null, [function () {
            try {
                return [clearable][0];
            } catch (e) {
                _e(e);
            }
        }.call(this) ? h('i', { 'ev-click': function () {
                try {
                    return [self._onClear][0];
                } catch (e) {
                    _e(e);
                }
            }.call(this) }, null, _className(function () {
            try {
                return [{ "k-clear ion-ios-close": true, "k-show": hasValue }][0];
            } catch (e) {
                _e(e);
            }
        }.call(this))) : undefined, h('i', null, null, 'k-arrow ion-arrow-down-b')], 'k-suffix')], 'k-wrapper', null, function (i) {
            widgets['wrapper'] = i;
        }), function () {
            try {
                return [Menu][0];
            } catch (e) {
                _e(e);
            }
        }.call(this)], '_context': $this }, null, null, null, function (i) {
        widgets['dropdown'] = i;
    }), _className(function () {
        try {
            return [(0, _extends3.default)({}, classNameObj, { 'k-has-value': hasValue })][0];
        } catch (e) {
            _e(e);
        }
    }.call(this)));
};
