// kpc/@css/components/spinner/index.js
'use strict';

exports.__esModule = true;
exports.Spinner = exports.default = undefined;

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _class, _temp;

var _intact = require('intact');

var _intact2 = _interopRequireDefault(_intact);

var _index = require('./index.vdt');

var _index2 = _interopRequireDefault(_index);

require('../../styles/kpc.css');

require('./index.css');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var numberReg = /^(-|\+)?\d+(\.(\d+)?)?$/;

var Spinner = (_temp = _class = function (_Intact) {
    (0, _inherits3.default)(Spinner, _Intact);

    function Spinner() {
        (0, _classCallCheck3.default)(this, Spinner);
        return (0, _possibleConstructorReturn3.default)(this, _Intact.apply(this, arguments));
    }

    Spinner.prototype.defaults = function defaults() {
        return {
            disabled: false,
            value: 0,
            max: Number.POSITIVE_INFINITY,
            min: Number.NEGATIVE_INFINITY,
            step: 1,
            size: 'default',
            vertical: false
        };
    };

    Spinner.prototype._init = function _init() {
        var _this2 = this;

        this._fixValue();
        this.on('$receive:value', this._fixValue);
        this.on('$change:_value', function (c, val) {
            var _get = _this2.get(),
                max = _get.max,
                min = _get.min;
            // if the _value is valid, then set it to value


            if (numberReg.test(val)) {
                val = Number(val);
                if (val <= max && val >= min) {
                    _this2.set('value', val);
                }
            }
        });
    };

    Spinner.prototype._fixValue = function _fixValue() {
        var value = this.get('value');
        if (value == null) {
            var min = this.get('min');
            if (min === Number.NEGATIVE_INFINITY) {
                value = 0;
            } else {
                value = min;
            }
        }
        this.set({
            '_value': value,
            'value': value
        });
        this.initValue = value;
    };

    Spinner.prototype._increase = function _increase(e) {
        var _get2 = this.get(),
            _value = _get2._value,
            step = _get2.step;

        this.set('_value', Number((_value + step).toFixed(10)));
    };

    Spinner.prototype._decrease = function _decrease(e) {
        var _get3 = this.get(),
            _value = _get3._value,
            step = _get3.step;

        this.set('_value', Number((_value - step).toFixed(10)));
    };

    Spinner.prototype._disableDecrease = function _disableDecrease() {
        var _get4 = this.get(),
            _value = _get4._value,
            min = _get4.min,
            step = _get4.step,
            disabled = _get4.disabled;

        return disabled || +_value <= min || Number((min + step).toFixed(10)) > _value;
    };

    Spinner.prototype._disableIncrease = function _disableIncrease() {
        var _get5 = this.get(),
            _value = _get5._value,
            max = _get5.max,
            step = _get5.step,
            disabled = _get5.disabled;

        return disabled || +_value >= max || Number((max - step).toFixed(10)) < _value;
    };

    Spinner.prototype._changeValue = function _changeValue(e) {
        var val = e.target.value.trim();

        var _get6 = this.get(),
            disabled = _get6.disabled,
            max = _get6.max,
            min = _get6.min;

        if (!numberReg.test(val) || disabled) {
            this.set('_value', this.initValue);
        } else {
            val = Number(val);
            if (val >= max) {
                val = max;
            } else if (val < min) {
                val = min;
            }
            this.set('_value', val);
        }
    };

    (0, _createClass3.default)(Spinner, [{
        key: 'template',
        get: function get() {
            return _index2.default;
        }
    }]);
    return Spinner;
}(_intact2.default), _class.propTypes = {
    disabled: Boolean,
    max: Number,
    min: Number,
    step: Number,
    vertical: Boolean
}, _temp);
exports.default = Spinner;
exports.Spinner = Spinner;
